Trigger build with .tex file watcher

harmishhk opened the pull request:
Adds a file watcher to trigger builds when external program changes a `.tex` file. Proposal for this feature is in Issue #151 .

James-Yu commented:
Thank you for your PR. This feature can be useful and will be merged after making the following changes
- The file extension check in `build()` should be respected. Suggested edits will be placed inline.
- A new configuration item, defaults to `false`, should be introduced to control whether auto-build on `tex` changes.

James-Yu commented on src/commander.ts:
```diff
@@ -14,9 +14,9 @@ export class Commander {
-        if (!vscode.window.activeTextEditor || !this.extension.manager.isTex(vscode.window.activeTextEditor.document.fileName)) {
-            return
-        }
+        // if (!vscode.window.activeTextEditor || !this.extension.manager.isTex(vscode.window.activeTextEditor.document.fileName)) {
```
This file extension check should be maintained. In the future I will examine its necessity. However it should not be in this PR.

James-Yu commented on src/main.ts:
```diff
@@ -152,6 +152,19 @@ export async function activate(context: vscode.ExtensionContext) {
+            return
+        }
+        if (extension.manager.isTex(e.fsPath)) {
+            extension.commander.build()
```
Instead of calling `build()` in `commander` (which is designed to be invoked upon vscode commands), you may use the same code from https://github.com/James-Yu/LaTeX-Workshop/pull/150/files#diff-c7f5bd618dbb862973c3204edf8120d6R20 to build the file.

harmishhk commented:
Thanks for reviewing the PR. I was also thinking of adding the configuration item for this functionality, which I have done in the [last commit](https://github.com/James-Yu/LaTeX-Workshop/pull/150/commits/9126ac8a228de11e32ee633c23d8f2d9e2ac07c8). Please make sure you like the wording `autoBuildOnTexChange` for this configuration option which defaults to `false`. I also added a note that this configuration option only works if `autoBuild` is set to `true`.

Also, it makes sense not to touch the `build()` function of `commander.ts`, so I have added the code inline.

Regarding efficiency/performance: Just adding the `FileWatcher` does not seem to affect the overall editor performance. I observed CPU usage of `code` processes with type `watcherService` using `htop` command on `Ubuntu 16.04`. These processes does not take any CPU when idle. Only when a `TeX` file changes that trigger the build the process uses the CPU until the build is finished. I was using this FileWatcher since last three updates of LaTeX-Workshop by rebasing my `file-watcher` branch on upstream `master`, and did not notice any performance issues with the editor. I am not aware of any other profiling method that could tell me exact impact of enabling a `FileWatcher`, do you have any suggestions?

If you still think the performance can be an issue, we can check the if the configuration option `autoBuildOnTexChange` is enabled even before creating the file watcher in `main.ts`. This, however, would require the user to reload the workspace for change in this particular configuration to take effect. I personally have no problem with that option either.

James-Yu commented:
It seems the `filewatcher` is not causing much trouble so I will leave it as is and let the users tell.

Thanks a lot for your contribution!
